### PR TITLE
tests/posix semaphore: fix test4

### PR DESCRIPTION
**Pull request by cladmi**
Fix posix_semaphore test4. The tests uses `sem_timedwait` to wait 1 second and verify it is really 1 second with 100 usec precision.

 However the test was measuring time including `puts` functions and so was late.
Now only `sem_timedwait` is measured.

Fix for `native`, as computers are not real time, it often happens that elapsed time is too big by more than 100 and sometime 200 us. So give native a bit more time here.

Refactoring and renaming in different commits.

### Question 

Why does the test allow a less than 1 second elapsed time?
Is it because `sem_timedwait` uses a different timer than `xtimer` and so could have 100us error margin?

### CODE REVIEW

File: tests/posix_semaphore/main.c [hunk 1]
```diff
@@ -260,13 +266,20 @@ void test4(void)
             puts("first: timed out");
         }
     }
-    stop = xtimer_now_usec64() - start;
-    if ((stop < (exp - 100)) || (stop > (exp + 100))) {
-        fmt_u64_dec(uint64_str, stop);
+
+    fmt_u64_dec(uint64_str, elapsed);
```
Comment (kaspar030) [inline tests/posix_semaphore/main.c]: fmt doesn't null terminate. either you save the length and supply it to printf, or you do sth like:
```uint64_str[fmt_u64_dec(uint64_str, elapsed)] = '\0';```

File: tests/posix_semaphore/main.c [hunk 2]
```diff
@@ -260,13 +266,20 @@ void test4(void)
...
+    if (elapsed < (exp - 100)) {
         printf("first: waited only %s usec => FAILED\n", uint64_str);
     }
+#ifdef BOARD_NATIVE
```
Comment (kaspar030) [inline tests/posix_semaphore/main.c]: Please define the literal somewhere "upstairs" and use that define in the logic.

File: tests/posix_semaphore/main.c (hunk 1 above)
Comment (cladmi) [reply tests/posix_semaphore/main.c]: The problem was also there before. But I will fix it.

File: tests/posix_semaphore/main.c (hunk 2 above)
Comment (cladmi) [reply tests/posix_semaphore/main.c]: Sure.

**Comment by miri64** [comment]
@cladmi ping?

**Comment by cladmi** [comment]
I addressed both changes. Please give ideas if you do not like the configuration variable name.
However, I think only the last 'fixup!' commit should be squashed, other ones are independent compiling changes.

**Comment by cladmi** [comment]
No ones has remarks regarding allowing a timeout less than the expected one ?

File: tests/posix_semaphore/main.c [hunk 3]
```diff
@@ -267,16 +274,11 @@ void test4(void)
         }
     }
 
-    fmt_u64_dec(uint64_str, elapsed);
+    uint64_str[fmt_u64_dec(uint64_str, elapsed)] = '\0';
     if (elapsed < (exp - 100)) {
```
Comment (kaspar030) [inline tests/posix_semaphore/main.c]: Maybe use the same margin here?

Comment (cladmi) [reply tests/posix_semaphore/main.c]: Why ? if `native` is slower, there is no reason to allow a bigger 'before' timeout.

File: tests/posix_semaphore/main.c [hunk 4]
```diff
@@ -260,13 +273,15 @@ void test4(void)
...
     }
-    stop = xtimer_now_usec64() - start;
-    if ((stop < (exp - 100)) || (stop > (exp + 100))) {
-        fmt_u64_dec(uint64_str, stop);
+
+    uint64_str[fmt_u64_dec(uint64_str, elapsed)] = '\0';
+    if (elapsed < (exp - 100)) {
         printf("first: waited only %s usec => FAILED\n", uint64_str);
     }
+    else if (elapsed > (exp + TEST4_TIMEOUT_EXCEEDED_MARGIN))
```
Comment (miri64) [inline tests/posix_semaphore/main.c]: `{` missing

Comment (cladmi) [reply tests/posix_semaphore/main.c]: My bad, I forgot to rerun 'static-test'.

**Comment by miri64** [comment]
Backport provided in #7808

**Pull request merged**